### PR TITLE
Issue 2426: Sushiswap

### LINKED ISSUE #2426: Add Sushiswap support
## Abstract

It seems that Sushiswap transactions are not currently picked up as trades in the profit/loss section of the application

## Motivation

Sushiswap is one of the primary decentralised exchanges and I would imagine that since it is a Uniswap fork, it should take less code to implement this feature than to implement tracking for an entirely different decentralised exchange

## Specification

N/A

### PR DESCRIPTION
**Pull request by yabirgb**
Closes #2426 

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

**Review by LefterisJP — COMMENTED** [review]
you got linting failures @yabirgb. Also some lgtm errors

File: rotkehlchen/db/upgrades/v24_v25.py [hunk 1]
```diff
@@ -469,7 +469,25 @@ def upgrade_v24_to_v25(db: 'DBHandler') -> None:
     cursor.execute(
         f'DELETE FROM used_query_ranges WHERE name LIKE "{BALANCER_EVENTS_PREFIX}%";',
     )
-    cursor.execute('DELETE FROM uniswap_events;')
+    cursor.execute('DROP TABLE IF EXISTS uniswap_events;')
```
Comment (yabirgb) [inline rotkehlchen/db/upgrades/v24_v25.py:472]: Everything should be ok now. There was an issue with this upgrade when coming from an older version (in the test for restoring the premium synced db), seems that the table was created using the initialization in the db but is not the case anymore since i deleted that sentence from the `schema.py` file.

File: rotkehlchen/chain/ethereum/modules/uniswap/uniswap.py [hunk 2]
```diff
@@ -1216,73 +834,53 @@ def get_balances(
...
         self,
         addresses: List[ChecksumEthAddress],
         reset_db_data: bool,
         from_timestamp: Timestamp,
         to_timestamp: Timestamp,
-    ) -> AddressEventsBalances:
-        """Get the addresses' events history in the Uniswap protocol
+    ) -> AddressTrades:
+        """Get the addresses' trades history in the Uniswap protocol
         """
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/uniswap/uniswap.py:845]:
```suggestion
        """Get the addresses' trades history in the Uniswap protocol"""
```

File: rotkehlchen/chain/ethereum/modules/uniswap/uniswap.py [hunk 3]
```diff
@@ -1216,73 +834,53 @@ def get_balances(
...
-        with self.trades_lock:
-            if reset_db_data is True:
-                self.database.delete_uniswap_trades_data()
-
-            trades = self._get_trades(
-                addresses=addresses,
-                from_timestamp=from_timestamp,
-                to_timestamp=to_timestamp,
-                only_cache=False,
+        """Fetch all DB Uniswap trades within the time range"""
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/uniswap/uniswap.py:865]: Can't this also be abstracted in the AMM swap parent class?

File: rotkehlchen/db/dbhandler.py [hunk 4]
```diff
@@ -1111,61 +1115,61 @@ def add_uniswap_events(self, events: Sequence[UniswapPoolEvent]) -> None:
                 cursor.execute(query, event_tuple)
             except sqlcipher.IntegrityError:  # pylint: disable=no-member
                 self.msg_aggregator.add_warning(
-                    f'Tried to add a Uniswap event that already exists in the DB. '
+                    f'Tried to add a AMM event that already exists in the DB. '
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:1118]:
```suggestion
                    f'Tried to add an AMM event that already exists in the DB. '
```

File: rotkehlchen/db/schema.py [hunk 5]
```diff
@@ -490,8 +492,8 @@
 );
 """
 
-DB_CREATE_UNISWAP_EVENTS = """
-CREATE TABLE IF NOT EXISTS uniswap_events (
+DB_CREATE_AMM_EVENTS = """
```
Comment (LefterisJP) [inline rotkehlchen/db/schema.py:495]: Don't do that. Make an issue to rename this in the next DB upgrade. Take these changes out of the PR.

As you can see, nothing else needs to change so there is no need to do a DB upgrade just for this. Use the old name **temporarily** and next time we do a DB upgrade, we also rename this. Make an issue to remind us.

File: rotkehlchen/db/upgrades/v24_v25.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/db/upgrades/v24_v25.py:472]: As I wrote above, don't do the upgrade now.

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+
+UNISWAP_TEST_OPTIONS = [
+    # Test with infura (as own node), many open nodes, and premium + graph
+    (False, INFURA_TEST, (NodeName.OWN, NodeName.MYCRYPTO)),
+    (False, '', (NodeName.MYCRYPTO, NodeName.BLOCKSCOUT, NodeName.AVADO_POOL)),
+    (True, '', ()),
+]
+# Skipped infura and many open nodes for now in the CI. Fails flakily due to timeouts
+# from time to time. We should run locally to make sure that it still works.
+SKIPPED_UNISWAP_TEST_OPTIONS = [UNISWAP_TEST_OPTIONS[-1]]
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:65]: In this file you got UNISWAP_TEST_OPTIONS.

Perhaps abstract it in some kind of utils test file we got and rename to AMM_TEST_OPTIONS and use both in this test file and uniswap one

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+
+
+@pytest.mark.parametrize('ethereum_accounts', [[SWAP_ADDRESS]])
+@pytest.mark.parametrize('ethereum_modules', [['uniswap']])
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:69]: Hm .... you are activating the uniswap module here, and not sushiswap.

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+    'start_with_valid_premium,ethrpc_endpoint,ethereum_manager_connect_at_start',
+    SKIPPED_UNISWAP_TEST_OPTIONS,
+)
+def test_get_balances(
+        rotkehlchen_api_server,
+        ethereum_accounts,  # pylint: disable=unused-argument
+        rotki_premium_credentials,
+        start_with_valid_premium,
+):
+    """Check querying the uniswap balances endpoint works. Uses real data
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:80]: Hmm ... man this entire test, tests uniswap. You use the uniswap module, call uniswal balance endpoint. Too much copy past :sweat_smile: 

Please change it to test sushiswap

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+    premium = None
+    if start_with_valid_premium:
+        premium = Premium(rotki_premium_credentials)
+
+    # Set module premium attribute
+    uniswap = rotki.chain_manager.get_module('uniswap')
+    uniswap.premium = premium
+
+    response = requests.get(
+        api_url_for(rotkehlchen_api_server, 'uniswapbalancesresource'),
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:99]: here is why this works, you test uniswap

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+            token1=EthereumToken('0xdAC17F958D2ee523a2206206994597C13D831ec7'),
+            amount0_in=AssetAmount(ZERO),
+            amount1_in=AssetAmount(FVal('506.399839')),
+            amount0_out=AssetAmount(FVal('1')),
+            amount1_out=AssetAmount(ZERO),
+        )],
+    )]
+
+
+def _query_and_assert_simple_uniswap_trades(setup, api_server, async_query):
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:355]:
```suggestion
def _query_and_assert_simple_sushiswap_trades(setup, api_server, async_query):
```

File: rotkehlchen/tests/api/test_sushiswap.py [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+@pytest.mark.parametrize('ethereum_accounts', [[SWAP_ADDRESS]])
+@pytest.mark.parametrize('ethereum_modules', [['sushiswap']])
+@pytest.mark.parametrize('start_with_valid_premium', [True])
+def test_get_sushiswap_trades_history(
+        rotkehlchen_api_server,
+        ethereum_accounts,  # pylint: disable=unused-argument
+        rotki_premium_credentials,  # pylint: disable=unused-argument
+        start_with_valid_premium,  # pylint: disable=unused-argument
+):
+    """Test that the last 11/23 uniswap trades of the account since 1605437542
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_sushiswap.py:387]:
```suggestion
    """Test that the last 11/23 sushiswap trades of the account since 1605437542
```

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 12]
```diff
@@ -2193,6 +2196,47 @@ def test_upgrade_db_27_to_28(user_data_dir):  # pylint: disable=unused-argument
     assert db.get_version() == 28
 
 
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+def test_upgrade_db_28_to_29(user_data_dir):  # pylint: disable=unused-argument
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2200]: as I said don't do an upgrade just for this

**Review by LefterisJP — COMMENTED** [review]
Thank you @yabirgb. Lots of comments in the first pass.

Please do all changes in new commits so it's easier for me to review.

File: rotkehlchen/chain/ethereum/modules/ammswap/ammswap.py [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+        in DB and finally all DB events are read, and processed for calculating
+        total profit/loss per LP (stored within <LiquidityPoolEventsBalance>).
+        """
+        raise NotImplementedError
+
+    # -- Methods following the EthereumModule interface -- #
+
+    @abc.abstractmethod
+    def deactivate(self) -> None:
+        raise NotImplementedError
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/ammswap/ammswap.py:561]:
```suggestion
    # -- Methods following the EthereumModule interface -- #

    @abc.abstractmethod
    def deactivate(self) -> None:
        raise NotImplementedError
```

Why is this needed? This is not a module. And this will conflict with the implementation of deactivate() in the other EthereumModule supercass since you do multiple inheritance. Don't do that.

Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/ammswap/ammswap.py:561]: It was because the other interface methods were the same and I wanted to get them here but it was not okey for the linter. And this one I didn't notice but is true that is the same in both and get overwritten. Going to remove it from here

File: rotkehlchen/tests/db/test_db_upgrades.py (hunk 12 above)
Comment (yabirgb) [reply rotkehlchen/tests/db/test_db_upgrades.py:2200]: What would you do here then?

File: rotkehlchen/chain/ethereum/modules/ammswap/ammswap.py [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+        addresses: List[ChecksumEthAddress],
+        reset_db_data: bool,
+        from_timestamp: Timestamp,
+        to_timestamp: Timestamp,
+    ) -> AddressEventsBalances:
+        """Get the addresses' events history in the Uniswap protocol
+        """
+        with self.trades_lock:
+            if reset_db_data is True:
+                self.database.delete_uniswap_events_data()
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/ammswap/ammswap.py:450]: That one I should have noticed. Sorry for that

File: rotkehlchen/chain/ethereum/modules/sushiswap/sushiswap.py [hunk 15]
```diff
@@ -0,0 +1,666 @@
...
+                result = self.graph.query(
+                    querystr=querystr,
+                    param_types=param_types,
+                    param_values=param_values,
+                )
+            except RemoteError as e:
+                self.msg_aggregator.add_error(SUBGRAPH_REMOTE_ERROR_MSG.format(error_msg=str(e)))
+                break
+
+            for entry in result['swaps']:
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/sushiswap/sushiswap.py:452]: We pray this never happens. But this is not handle neither here or uniswap and it's needed :heavy_check_mark:

File: rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py [hunk 16]
```diff
@@ -0,0 +1,991 @@
...
+            unknown_asset_price: AssetToPrice,
+    ) -> None:
+        """Update the pools underlying assets prices in USD (prices obtained
+        via Inquirer and the subgraph)
+        """
+        for lps in address_balances.values():
+            for lp in lps:
+                # Try to get price from either known or unknown asset price.
+                # Otherwise keep existing price (zero)
+                total_user_balance = FVal(0)
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py:326]: nitpick: Instead of `FVal(0)` use the `ZERO` constant defined in constants/misc.py

File: rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py [hunk 17]
```diff
@@ -0,0 +1,991 @@
...
+        """
+        address_amm_trades: AddressTrades = {}
+        new_addresses: List[ChecksumEthAddress] = []
+        existing_addresses: List[ChecksumEthAddress] = []
+        min_end_ts: Timestamp = to_timestamp
+
+        if only_cache:
+            return self._fetch_trades_from_db(addresses, from_timestamp, to_timestamp)
+
+        # Get addresses' last used query range for Sushiswap trades
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py:638]: Comment references sushiswap

```suggestion
        # Get addresses' last used query range for this AMM's trades
```

File: rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py [hunk 18]
```diff
@@ -0,0 +1,991 @@
...
+            if db_trades:
+                db_address_trades[address] = db_trades
+
+        return db_address_trades
+
+    def _get_balances_graph(
+            self,
+            addresses: List[ChecksumEthAddress],
+    ) -> ProtocolBalance:
+        """Get the addresses' pools data querying the Sushiswap subgraph
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py:833]: Sushiswap should go from here.
```suggestion
        """Get the addresses' pools data querying the AMM's subgraph
```

File: rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py [hunk 19]
```diff
@@ -0,0 +1,991 @@
...
+            for lp in result_data:
+                lp_pair = lp['pair']
+                lp_total_supply = FVal(lp_pair['totalSupply'])
+                user_lp_balance = FVal(lp['liquidityTokenBalance'])
+                try:
+                    user_address = deserialize_ethereum_address(lp['user']['id'])
+                    lp_address = deserialize_ethereum_address(lp_pair['id'])
+                except DeserializationError as e:
+                    msg = (
+                        f'Failed to Deserialize address. Skipping pool {lp_pair} '
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py:881]: need to add location here
```suggestion
                        f'Failed to Deserialize address. Skipping {self.location} pool {lp_pair} '
```

File: rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py [hunk 20]
```diff
@@ -0,0 +1,991 @@
...
+                token1['total_amount'] = lp_pair['reserve1']
+
+                liquidity_pool_assets = []
+                for token in token0, token1:
+                    try:
+                        deserialized_eth_address = deserialize_ethereum_address(token['id'])
+                    except DeserializationError as e:
+                        msg = (
+                            f'Failed to deserialize token address {token["id"]} '
+                            f'Bad token address in lp pair came from the graph.'
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/interfaces/ammswap/ammswap.py:900]: location here
```suggestion
                            f'Bad token address in {self.location} lp pair came from the graph.'
```

**Review by LefterisJP — APPROVED** [review]
LGTM. Left some nitpick comments which I will handle in another small PR that I wanted to do. Will merge despite them since it's big, I can't find anything wrong with a quick extra look and all is green.

What I would like you to do is:
1. Make an issue and schedule for 1.21 to do include renaming of the table as requested here: https://github.com/rotki/rotki/pull/3259#discussion_r679013085
2. Make a small PR adding the 3 new endpoints to the api docs.
